File: Difficulty: Easy/Hashing for pair - 2/hashing-for-pair-2.java
//{ Driver Code Starts
// Initial Template for Java

// Initial Template for Java

import java.io.*;
import java.util.*;

class Gfg {
    public static void main(String[] args) {

        // Taking input using class Scanner
        Scanner sc = new Scanner(System.in);

        // Taking total count of testcases
        int testcases = Integer.parseInt(sc.nextLine());
        while (testcases-- > 0) {
            // taking size of array
            String temp[] = sc.nextLine().trim().split(" ");
            int N = temp.length;
            int arr[] = new int[N]; // array declaration

            for (int i = 0; i < N; i++)
                arr[i] = Integer.parseInt(temp[i]); // Input the
                                                    // array

            int sum = 0;

            // taking sum
            sum = Integer.parseInt(sc.nextLine());
            // Creating an object of class Geeks
            Geeks obj = new Geeks();

            // Calling sumExists method of class Geeks
            // and printing the result
            System.out.println(obj.sumExists(arr, sum));

            System.out.println("~");
        }
    }
}


// } Driver Code Ends

class Geeks {
    // Function to check if two numbers in the array have a sum equal to the given sum
    public static int sumExists(int arr[], int sum) {
        // Use a LinkedHashSet to store elements and check for the complement of each element
        LinkedHashSet<Integer> al = new LinkedHashSet<>();
        
        // Iterate through the array to find if any two numbers sum up to the target sum
        for (int i : arr) {
            // Calculate the complement that, when added to the current element, equals the target sum
            int diff = sum - i;
            
            // If the complement already exists in the set, we found the pair
            if (al.contains(diff)) {
                return 1;  // Return 1 if a pair exists
            }
            
            // Add the current element to the set
            al.add(i);
        }
        
        // Return 0 if no such pair exists
        return 0;
    }
}



//{ Driver Code Starts.

// } Driver Code Ends

File: Difficulty: Easy/Hashing for pair - 2/hashing-for-pair-2.js
//{ Driver Code Starts
// Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(
        string => { return string.trim(); });

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    for (let i = 0; i < t; i++) {
        let input_line = readLine().split(' ');
        let N = parseInt(input_line[0]);

        let arr = new Array(N);

        input_line = readLine().split(' ');
        for (let i = 0; i < N; i++) {
            arr[i] = parseInt(input_line[i]);
        }

        input_line = readLine().split(' ');
        let sum = input_line[0];

        let obj = new Solution();
        let ans = obj.sumExists(arr, N, sum);
        if (ans === -0) ans = 0;
        console.log(ans);
    
console.log("~");
}
}

// } Driver Code Ends

// User function Template for JavaScript

/**
 * @param {number[]} arr - Array of numbers
 * @param {number} N - Length of the array
 * @param {number} sum - Target sum to check for
 * @returns {number} - Returns 1 if a pair with the given sum exists, otherwise 0
 */
class Solution {
    // Function to check if there exists a pair of numbers in the array
    // whose sum equals the given target sum.
    sumExists(arr, N, sum) {
        // Map to store required complements to achieve the target sum
        let requiredComplements = new Map();

        for (let i = 0; i < N; i++) {
            // If current number is already in the map, a valid pair exists
            if (requiredComplements.has(arr[i])) {
                return 1; // Pair found
            }
            // Store the complement required to form the sum with current number
            requiredComplements.set(sum - arr[i], arr[i]);
        }

        return 0; // No valid pair found
    }
}
